File: Strings/trie_datastructure.java
class Trie {
    class Ctrie{
        char alpha;
        int eol;
        Ctrie[] chartie;
        
        public Ctrie(char val, int end)
        {
            alpha = val;
            eol = end;
            chartie = new Ctrie[26];
        }
    }

    /** Initialize your data structure here. */
    Ctrie dummy;
    public Trie() {
        dummy = new Ctrie(' ',0);
    }
    
    /** Inserts a word into the trie. */
    public void insert(String word) {
        int n = word.length();
        Ctrie cur = dummy;
        for(int i=0;i<n;i++)
        {
            int pos = word.charAt(i)-'a';
            if(i!=n-1)
            {
                if(cur.chartie[pos]==null)
                {
                    cur.chartie[pos] = new Ctrie(word.charAt(i),0);
                }
                
                cur = cur.chartie[pos];
            }
            else{
                if(cur.chartie[pos]!=null && cur.chartie[pos].eol!=1)
                {
                    cur = cur.chartie[pos];
                    cur.eol=1;
                }
                else if(cur.chartie[pos]==null)
                {
                    cur.chartie[pos] = new Ctrie(word.charAt(i),1);
                }
                
            }
        }
        return;
    }
    
    /** Returns if the word is in the trie. */
    public boolean search(String word) {
        int n = word.length();
        Ctrie cur = dummy;
        for(int i=0;i<n;i++)
        {
            int pos = word.charAt(i)-'a';
            cur = cur.chartie[pos];
            if(i!=n-1)
            {
                if(cur==null)
                {
                    return false;
                }                                  
            }
            else{
               if(cur==null)
                {
                    return false;
                }                                  
                if(cur.eol!=1)
                {
                    return false;
                }  
            }
        }
        return true;
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    public boolean startsWith(String prefix) {
        int n = prefix.length();
        Ctrie cur = dummy;
        for(int i=0;i<n;i++)
        {
            int pos = prefix.charAt(i)-'a';
            cur = cur.chartie[pos];
            if(cur==null)
            {
                return false;
            }
        }
        return true;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */

File: Strings/trie_datastructure.py
class ctrie:
    def __init__(self,char,end):
        self.char = char
        self.eol = end
        self.ctriearr = [None]*26

class Trie:    

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.dummy = ctrie(' ',0)
        

    def insert(self, word: str) -> None:
        """
        Inserts a word into the trie.
        """
        n = len(word)
        cur = self.dummy
        i=0
        for char in word:
            pos = ord(char) - ord('a')
            if i!=n-1:
                if cur.ctriearr[pos]==None:
                    cur.ctriearr[pos] = ctrie(char,0)
                    #print(cur.ctriearr[pos].char)
                cur = cur.ctriearr[pos]
                i+=1
            else:
                if cur.ctriearr[pos]!=None:
                    if cur.ctriearr[pos].eol!=1:
                        cur.ctriearr[pos].eol=1
                elif cur.ctriearr[pos]==None:
                    cur.ctriearr[pos] = ctrie(char,1)
        return
                
            

    def search(self, word: str) -> bool:
        """
        Returns if the word is in the trie.
        """
        # listw = list(word)
        # print(listw)
        n = len(word)
        cur = self.dummy
        i=0
        for char in word:
            pos = ord(char) - ord('a')
            cur = cur.ctriearr[pos]
            if i!=n-1:
                #print(cur.char)
                if cur==None:
                    return False
                i+=1
            else:
                if cur!=None:
                    if cur.eol==0:
                        return False     
                if cur==None:
                    return False
                    
        return True
                

    def startsWith(self, prefix: str) -> bool:
        """
        Returns if there is any word in the trie that starts with the given prefix.
        """
        n = len(prefix)
        cur = self.dummy
        i=0
        for char in prefix:
            pos = ord(char) - ord('a')
            cur = cur.ctriearr[pos]
            if cur==None:
                return False
        return True
            
        
        


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)
